components/forwarder/forwarder_test.go

```diff
@@ -0,0 +1,176 @@
+package forwarder_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/components/forwarder"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/ThreeDotsLabs/watermill/pubsub/gochannel"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	logger = watermill.NewStdLogger(true, true)
+
+	forwarderTopic = "forwarder_topic"
+	outTopic       = "out_topic"
+)
+
+// TestForwarder tests forwarding messages from PubSubIn to PubSubOut (which are GoChannel implementation underneath).
+func TestForwarder(t *testing.T) {
+	// Create a set of publisher and subscribers for both In and Out Pub/Subs.
+	publisherIn, subscriberIn := newPubSubIn(logger)
+	publisherOut, subscriberOut := newPubSubOut(logger)
+	defer func() {
+		require.NoError(t, publisherIn.Close())
+		require.NoError(t, subscriberIn.Close())
+		require.NoError(t, publisherOut.Close())
+		require.NoError(t, subscriberOut.Close())
+	}()
+
+	// Create test context with a 5 seconds timeout so it will close any subscriptions/handlers running in the background
+	// in case of too long test execution.
+	ctx, cancelCtx := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancelCtx()
+
+	messageForwardedDetector, messageForwardedCh := setupMessageForwardedDetectorMiddleware()
```

Comment:
maybe using GoChannel Pub/Sub will be a bit better test? it will be more "end to end"

Reply:
so in other words, it may work like `gochannel Pub/Sub 1 -> forwarder -> gochannel Pub/Sub 2`

Reply:
Actually it's already like that. `newPubSubOut/In` create GoChannel Pub/Subs. Maybe it's misleading that I need this middleware. The middleware is helpful when I want to detect if the message was nacked (to not wait some fixed duration for NOT receiving the message on out topic.

Reply:
ah right, hmm I still feel lack of checking what "end subscriber" receive, so let's imagine that we will use router in the wrong way and message will be not forwarded properly after middleware

Reply:
ef8223a